Fix implementation for the built-in function '__builtin_alloca'

lucasccordeiro opened the pull request:
This PR should fix the issue reported in https://github.com/esbmc/esbmc/issues/330.

lucasccordeiro committed:
Added test cases for __builtin_alloca

Added test cases to check the pre- and post-conditions
of the builtin function __builtin_alloca.

lucasccordeiro commented:
@rafaelsamenezes: could you run this PR over the SV-COMP benchmarks?

In particular, we need to:

(1) Update the SV-COMP scripts to consider the FALSE result for the memory-safety category.
(2) Run the script from (1) using the master branch.
(3) Run the script from (1) using this PR.

mikhailramalho commented:
As I said before, simply removing the code is not the right approach. 

By removing this code, ESBMC does not verify the following program correctly: 
```
int main(argc, argv) char **argv;
{
  int *foo = __builtin_alloca(sizeof(int));
  free(foo);
}
```

lucasccordeiro commented:
Let's use this PR to fix this implementation of the built-in function `__builtin_alloca`.

lucasccordeiro committed:
Check dynamic object allocated using alloca

Check whether the object allocated with alloca is the same
as given in the free function.

lucasccordeiro committed:
Added Juliet test cases from SV-COMP

Signed-off-by: Lucas Cordeiro <lucasccordeiro@gmail.com>

mikhailramalho commented on regression/esbmc/github_330_4/test.desc:
```diff
@@ -0,0 +1,4 @@
+CORE
+main.c
+
```
Did you forget the `--memory-leak-check` here?

lucasccordeiro replied:
Fixed.

mikhailramalho reviewed (changes requested):
I'm almost fine with this PR, but I'd like to run on SV-COMP benchmarks before approving it.

The memory safety category only.

lucasccordeiro commented:
The results for `memory-safety` using this PR (with `incremental-bmc`). We have also updated the script to consider all TRUE and FALSE results.

```
Statistics:           3436 Files
  correct:            2716
    correct true:     1284
    correct false:    1432
  incorrect:           124
    incorrect true:      3
    incorrect false:   121
  unknown:             596
  Score:              1968 (max: 5231)
```

The results for `memory-safety` from SV-COMP 2021 (I just copied them from https://sv-comp.sosy-lab.org/2021/results/results-verified/). These results ignore FALSE.

```
Statistics:           3436 Files
  correct:            1224
    correct true:     1224
    correct false:    0
  incorrect:           3
    incorrect true:      3
    incorrect false:   0
  Score:              1281 (max: 5231)
```

The results for `memory-safety` using the master (with `incremental-bmc`). We have also updated the script to consider all TRUE and FALSE results.
```
Statistics:           3436 Files
  correct:            2574
    correct true:     1139
    correct false:    1435
  incorrect:           269
    incorrect true:      3
    incorrect false:   266
  unknown:             593
  Score:              -639 (max: 5231)
```

lucasccordeiro committed:
Added memory leak check for regression test

Signed-off-by: Lucas Cordeiro <lucasccordeiro@gmail.com>

lucasccordeiro commented:
I'm now waiting for the results using the master branch; I hope to have them by the end of the day.

mikhailramalho commented:
These svcomp results are using incremental approach without ignoring any
esbmc result, right?

On Wed, 13 Jan 2021, 08:51 Lucas Cordeiro, <notifications@github.com> wrote:

> I'm now waiting for the results using the master branch; I hope to have
> them by the end of the day.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/esbmc/esbmc/pull/337#issuecomment-759428426>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAKEJH2A6J4P5AUTEZPLAATSZWJL7ANCNFSM4V53ICDQ>
> .
>

lucasccordeiro commented:
No. I just copied them from the SV-COMP 2021 website.

Em qua., 13 de jan. de 2021 às 12:54, Mikhail R. Gadelha <
notifications@github.com> escreveu:

> These svcomp results are using incremental approach without ignoring any
> esbmc result, right?
>
> On Wed, 13 Jan 2021, 08:51 Lucas Cordeiro, <notifications@github.com>
> wrote:
>
> > I'm now waiting for the results using the master branch; I hope to have
> > them by the end of the day.
> >
> > —
> > You are receiving this because you commented.
> > Reply to this email directly, view it on GitHub
> > <https://github.com/esbmc/esbmc/pull/337#issuecomment-759428426>, or
> > unsubscribe
> > <
> https://github.com/notifications/unsubscribe-auth/AAKEJH2A6J4P5AUTEZPLAATSZWJL7ANCNFSM4V53ICDQ
> >
> > .
> >
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/esbmc/esbmc/pull/337#issuecomment-759429919>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AA4F4HPTI3QKZHB7FLNKQHTSZWJW7ANCNFSM4V53ICDQ>
> .
>


-- 
Lucas C. Cordeiro​

https://ssvlab.github.io/lucasccordeiro/

lucasccordeiro committed:
Merge branch 'master' into fix-alloca

lucasccordeiro commented:
@mikhailramalho and @rafaelsamenezes: This PR increases our scores significantly from -639 to 1968.

mikhailramalho reviewed (approved):
LGTM.

rafaelsamenezes reviewed (approved):
LGTM

After this is merged we should open another issue to make `__builtin_alloca` add into `stack`
